File: javascript/packages/linter/src/rules/erb-no-output-in-attribute-position.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"
import { isTagAttributesCall, isConditionalTagAttributesCall } from "./action-view-utils.js"
import { isERBNode, isERBOutputNode, isERBContentNode } from "@herb-tools/core"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ParseResult, HTMLOpenTagNode, ParserOptions } from "@herb-tools/core"

class ERBNoOutputInAttributePositionVisitor extends BaseRuleVisitor {
  visitHTMLOpenTagNode(node: HTMLOpenTagNode): void {
    for (const child of node.children) {
      if (!isERBNode(child)) continue
      if (!isERBOutputNode(child)) continue
      if (!isERBContentNode(child)) continue

      const prismNode = child.prismNode
      if (prismNode && isTagAttributesCall(prismNode)) continue

      if (prismNode && isConditionalTagAttributesCall(prismNode)) {
        this.addOffense(
          "Avoid using conditional `tag.attributes` in attribute position. Use `<% if ... %><%= tag.attributes(...) %><% end %>` instead.",
          child.location,
        )

        continue
      }

      this.addOffense(
        "Avoid `<%= %>` in attribute position. Use `<% if ... %>` with static attributes instead.",
        child.location,
      )
    }

    super.visitHTMLOpenTagNode(node)
  }
}

export class ERBNoOutputInAttributePositionRule extends ParserRule {
  static ruleName = "erb-no-output-in-attribute-position"
  static introducedIn = this.version("0.9.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      prism_nodes: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new ERBNoOutputInAttributePositionVisitor(this.ruleName, context)
    visitor.visit(result.value)
    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/erb_no_output_in_attribute_position.rs
use crate::utils::action_view_utils::{is_conditional_tag_attributes_call, is_tag_attributes_call};
use crate::utils::erb_utils::is_output_tag_opening;

use herb::nodes::{AnyNode, HTMLOpenTagNode};
use herb::Visitor;

rule_visitor!(ERBNoOutputInAttributePositionVisitor);
define_parser_rule!(
  ERBNoOutputInAttributePositionRule,
  "erb-no-output-in-attribute-position",
  Error,
  ERBNoOutputInAttributePositionVisitor,
  parser_options: { prism_nodes: true },
  introduced_in: "0.9.0"
);

impl Visitor for ERBNoOutputInAttributePositionVisitor {
  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    for child in &node.children {
      let erb = match child {
        AnyNode::ERBContentNode(erb) => erb,
        _ => continue,
      };

      let tag_opening = erb.tag_opening.as_ref().map(|token| token.value.as_str()).unwrap_or("");

      if !is_output_tag_opening(tag_opening) {
        continue;
      }

      if let Some(ref prism_node) = erb.prism_node_ast {
        if is_tag_attributes_call(prism_node) {
          continue;
        }

        if is_conditional_tag_attributes_call(prism_node) {
          self.add_offense(
            "Avoid using conditional `tag.attributes` in attribute position. Use `<% if ... %><%= tag.attributes(...) %><% end %>` instead.",
            erb.location.clone(),
          );

          continue;
        }
      }

      self.add_offense(
        "Avoid `<%= %>` in attribute position. Use `<% if ... %>` with static attributes instead.",
        erb.location.clone(),
      );
    }

    self.walk_html_open_tag_node(node);
  }
}
